### ISSUE TITLE
Fail to build on CentOS 7

### ISSUE THREAD

**Issue by Invincibear** [facebook/folly#915]
I seem to be stuck when making the installer
```
[root@fc38 folly]# make
make  all-recursive
make[1]: Entering directory `/root/folly/folly'
Making all in .
make[2]: Entering directory `/root/folly/folly'
depbase=`echo ClockGettimeWrappers.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
/bin/sh ./libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H   -I./.. -pthread -I/usr/include     -std=gnu++1y -g -O2 -MT ClockGettimeWrappers.lo -MD -MP -MF $depbase.Tpo -c -o ClockGettimeWrappers.lo ClockGettimeWrappers.cpp &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I./.. -pthread -I/usr/include -std=gnu++1y -g -O2 -MT ClockGettimeWrappers.lo -MD -MP -MF .deps/ClockGettimeWrappers.Tpo -c ClockGettimeWrappers.cpp  -fPIC -DPIC -o .libs/ClockGettimeWrappers.o
In file included from ./../folly/portability/Time.h:22:0,
                 from ./../folly/ClockGettimeWrappers.h:19,
                 from ClockGettimeWrappers.cpp:17:
./../folly/portability/Config.h:27:18: error: missing binary operator before token "("
 #if __has_include(<features.h>)
                  ^
make[2]: *** [ClockGettimeWrappers.lo] Error 1
make[2]: Leaving directory `/root/folly/folly'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/root/folly/folly'
make: *** [all] Error 2
```

This is what I've done up to this point, after installing cmake, cmake3, and all of the dependencies listed at the bottom of https://github.com/facebook/folly, including double-decimal

```
[root@fc38 ~]# pwd
/root


[root@fc38 ~]# git clone https://github.com/facebook/folly.git
Cloning into 'folly'...
remote: Counting objects: 48603, done.
remote: Compressing objects: 100% (111/111), done.
remote: Total 48603 (delta 94), reused 81 (delta 46), pack-reused 48446
Receiving objects: 100% (48603/48603), 18.77 MiB | 8.39 MiB/s, done.
Resolving deltas: 100% (39267/39267), done.


[root@fc38 ~]# cd folly/folly


[root@fc38 folly]# autoreconf -ivf
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4
autoreconf: configure.ac: tracing
autoreconf: configure.ac: creating directory build-aux
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, `build-aux'.
libtoolize: copying file `build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIR, `m4'.
libtoolize: copying file `m4/libtool.m4'
libtoolize: copying file `m4/ltoptions.m4'
libtoolize: copying file `m4/ltsugar.m4'
libtoolize: copying file `m4/ltversion.m4'
libtoolize: copying file `m4/lt~obsolete.m4'
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:32: installing 'build-aux/config.guess'
configure.ac:32: installing 'build-aux/config.sub'
configure.ac:27: installing 'build-aux/install-sh'
configure.ac:27: installing 'build-aux/missing'
Makefile.am: installing 'build-aux/depcomp'
parallel-tests: installing 'build-aux/test-driver'
test/Makefile.am:388: warning: variable 'detail_iterators_test_SOURCES' is defined but no program or
test/Makefile.am:388: library has 'detail_iterators_test' as canonical name (possible typo)
test/Makefile.am:389: warning: variable 'detail_iterators_test_LDADD' is defined but no program or
test/Makefile.am:389: library has 'detail_iterators_test' as canonical name (possible typo)
autoreconf: Leaving directory `.'


[root@fc38 folly]# ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for style of include used by make... GNU
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking dependency style of gcc... gcc3
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking dependency style of g++... gcc3
checking how to run the C++ preprocessor... g++ -E
checking for ld used by g++... /usr/bin/ld -m elf_x86_64
checking if the linker (/usr/bin/ld -m elf_x86_64) is GNU ld... yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC -DPIC
checking if g++ PIC flag -fPIC -DPIC works... yes
checking if g++ static flag -static works... no
checking if g++ supports -c -o file.o... yes
checking if g++ supports -c -o file.o... (cached) yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... (cached) GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking dependency style of gcc... (cached) gcc3
checking if g++ supports C++1y features without additional flags... no
checking if g++ supports C++1y features with -std=c++1y... yes
checking if g++ supports C++1y features with -std=gnu++1y... yes
checking whether -Wunknown-warning-option is supported... no
checking whether -Wshadow-local and -Wshadow-compatible-local are supported... no
checking for main in -lgflags... yes
checking for gflags viability... yes
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for GFLAGS... no
configure: gflags.pc not found, treating as legacy dependency
checking for main in -lglog... yes
checking for glog viability... yes
checking for GLOG... yes
checking for pkg-config... /usr/bin/pkg-config
checking whether compiling and linking against OpenSSL works... yes
checking for OPENSSL... yes
checking for boostlib >= 1.51.0... yes
checking whether the Boost::Context library is available... yes
checking for exit in -lboost_context-mt... yes
checking whether the Boost::Program_Options library is available... yes
checking for exit in -lboost_program_options-mt... yes
checking whether the Boost::Thread library is available... yes
checking for exit in -lboost_thread... yes
checking whether the Boost::Regex library is available... yes
checking for exit in -lboost_regex-mt... yes
checking whether the Boost::System library is available... yes
checking for exit in -lboost_system... yes
checking whether the Boost::Filesystem library is available... yes
checking for exit in -lboost_filesystem... yes
checking whether the Boost::Chrono library is available... yes
checking for exit in -lboost_chrono-mt... yes
checking for python... /usr/bin/python
checking for python version... 2.7
checking for python platform... linux2
checking for python script directory... ${prefix}/lib/python2.7/site-packages
checking for python extension module directory... ${exec_prefix}/lib64/python2.7/site-packages
checking for ANSI C header files... (cached) yes
checking features.h usability... yes
checking features.h presence... yes
checking for features.h... yes
checking malloc.h usability... yes
checking malloc.h presence... yes
checking for malloc.h... yes
checking bits/functexcept.h usability... yes
checking bits/functexcept.h presence... yes
checking for bits/functexcept.h... yes
checking bits/c++config.h usability... yes
checking bits/c++config.h presence... yes
checking for bits/c++config.h... yes
checking double-conversion/double-conversion.h usability... yes
checking double-conversion/double-conversion.h presence... yes
checking for double-conversion/double-conversion.h... yes
checking for ceil in -ldouble-conversion... yes
checking for event_set in -levent... yes
checking for EVENT... yes
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking for size_t... yes
checking whether time.h and sys/time.h may both be included... yes
checking for working volatile... yes
checking for __int128... yes
checking for __int128 type traits... yes
checking for pthread_spinlock_t... yes
checking for ifunc support... yes
checking for final and override support... yes
checking for libc++... no
checking for c++11 atomic support without GNU Atomic library... yes
checking for main in -ldl... yes
checking for liblinux-vdso support... yes
checking for usable std::is_trivially_copyable... no
checking gflags namespace... gflags
checking for weak symbol support... yes
checking for wchar support... yes
checking for unaligned access support... yes
checking for vsnprintf reporting bad format strings... no
checking for library containing cplus_demangle_v3_callback... no
checking for library containing clock_gettime... none required
checking for pthread_atfork support... yes
checking for XSI style strerror_r support... no
checking for ext/random and __gnu_cxx::sfmt19937... yes
checking for VLA (variable-length array) support... yes
checking for variable template support... no
checking for malloc_size... no
checking for malloc_usable_size... yes
checking for memrchr... yes
checking for pipe2... yes
checking for preadv... yes
checking for pwritev... yes
checking lz4.h usability... yes
checking lz4.h presence... yes
checking for lz4.h... yes
checking for LZ4_decompress_safe in -llz4... yes
checking snappy.h usability... yes
checking snappy.h presence... yes
checking for snappy.h... yes
checking for main in -lsnappy... yes
checking zlib.h usability... yes
checking zlib.h presence... yes
checking for zlib.h... yes
checking for main in -lz... yes
checking lzma.h usability... no
checking lzma.h presence... no
checking for lzma.h... no
checking zstd.h usability... no
checking zstd.h presence... no
checking for zstd.h... no
checking bzlib.h usability... no
checking bzlib.h presence... no
checking for bzlib.h... no
checking linux/membarrier.h usability... no
checking linux/membarrier.h presence... no
checking for linux/membarrier.h... no
checking libdwarf/dwarf.h usability... yes
checking libdwarf/dwarf.h presence... yes
checking for libdwarf/dwarf.h... yes
checking dwarf.h usability... no
checking dwarf.h presence... no
checking for dwarf.h... no
checking elf.h usability... yes
checking elf.h presence... yes
checking for elf.h... yes
checking for backtrace in -lunwind... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating chrono/test/Makefile
config.status: creating io/test/Makefile
config.status: creating libfolly.pc
config.status: creating test/Makefile
config.status: creating test/function_benchmark/Makefile
config.status: creating experimental/Makefile
config.status: creating experimental/io/test/Makefile
config.status: creating experimental/symbolizer/Makefile
config.status: creating init/Makefile
config.status: creating logging/example/Makefile
config.status: creating stats/test/Makefile
config.status: creating config.h
config.status: executing folly-config.h commands
config.status: creating folly-config.h - prefix FOLLY for config.h defines
config.status: executing depfiles commands
config.status: executing libtool commands
```

**Comment by yfeldblum** [comment]
Which compiler are you using?

**Comment by daniel-lucio** [comment]
I had the same issue, I fixed by updating GCC. It works under 5.3.1.

I will publish soon an RPM for that. Just be patient.

**Comment by yfeldblum** [comment]
Right, `__has_include` is available for gcc >= 4.9.2. It is not available in gcc <= 4.9.1, and folly also does not work on gcc <= 4.9.1.

**Comment by daniel-lucio** [comment]
RPM published, check #929
